Userland/Utilities/tail.cpp

```diff
@@ -5,67 +5,42 @@
  */
 
 #include <LibCore/ArgsParser.h>
-#include <LibCore/File.h>
+#include <LibCore/EventLoop.h>
+#include <LibCore/FileWatcher.h>
+#include <LibCore/Stream.h>
 #include <LibCore/System.h>
-#include <LibMain/Main.h>
-#include <stdlib.h>
-#include <unistd.h>
 
 #define DEFAULT_LINE_COUNT 10
 
-static int tail_from_pos(Core::File& file, off_t startline, bool want_follow)
+static ErrorOr<void> tail_from_pos(Core::Stream::File& file, off_t startline)
 {
-    if (!file.seek(startline + 1))
-        return 1;
-
-    while (true) {
-        auto const& b = file.read(4096);
-        if (b.is_empty()) {
-            if (!want_follow) {
-                break;
-            } else {
-                while (!file.can_read()) {
-                    // FIXME: would be nice to have access to can_read_from_fd with an infinite timeout
-                    usleep(100);
-                }
-                continue;
-            }
-        }
-
-        if (write(STDOUT_FILENO, b.data(), b.size()) < 0)
-            return 1;
-    }
-
-    return 0;
+    TRY(file.seek(startline + 1, Core::Stream::SeekMode::SetPosition));
+    out(TRY(file.read_all()));
+    return {};
 }
 
-static off_t find_seek_pos(Core::File& file, int wanted_lines)
+static ErrorOr<off_t> find_seek_pos(Core::Stream::File& file, int wanted_lines)
 {
     // Rather than reading the whole file, start at the end and work backwards,
     // stopping when we've found the number of lines we want.
-    off_t pos = 0;
-    if (!file.seek(0, Core::SeekMode::FromEndPosition, &pos)) {
-        warnln("Failed to find end of file: {}", file.error_string());
-        return 1;
-    }
+    off_t pos = TRY(file.seek(0, Core::Stream::SeekMode::FromEndPosition));
 
     off_t end = pos;
     int lines = 0;
 
-    // FIXME: Reading char-by-char is only OK if IODevice's read buffer
-    // is smart enough to not read char-by-char. Fix it there, or fix it here :)
     for (; pos >= 0; pos--) {
-        file.seek(pos);
-        auto const& ch = file.read(1);
-        if (ch.is_empty()) {
-            // Presumably the file got truncated?
-            // Keep trying to read backwards...
-        } else {
+        TRY(file.seek(pos, Core::Stream::SeekMode::SetPosition));
+
+        if (!file.is_eof()) {
+            Array<u8, 1> buffer;
+            auto ch = TRY(file.read(buffer));
             if (*ch.data() == '\n' && (end - pos) > 1) {
                 lines++;
                 if (lines == wanted_lines)
                     break;
             }
+        } else {
+            break;
```

Comment:
`if (file.is_eof()) break;`, so that the rest of the code doesn't have to be indented as much.